File: 05-Algorithms/code/python/seb_alg/zero_one_bfs.py
"""Mirrors `typescript/src/zeroOneBfs.ts` exactly."""

from __future__ import annotations

import math
from collections import deque
from typing import List, Optional, Sequence

from .errors import AlgoError
from .graph_common import build_weighted_adj_list, check_vertex_in_range


def zero_one_bfs(n: int, edges: Sequence[Sequence[int]], src: int, directed: bool = False) -> List[Optional[int]]:
    """0-1 BFS via deque (0-weight edges pushed front, 1-weight pushed back).
    O(V + E). Requires weights in {0, 1}. Unreachable vertices are None.
    """
    check_vertex_in_range(n, src)
    for _, _, w in edges:
        if w != 0 and w != 1:
            raise AlgoError("invalid", "zeroOneBfs requires weights in {0, 1}")
    adj = build_weighted_adj_list(n, edges, directed)
    dist: List[float] = [math.inf] * n
    dist[src] = 0
    dq: deque = deque([src])
    visited = [False] * n

    while dq:
        u = dq.popleft()
        if visited[u]:
            continue
        visited[u] = True
        for v, w in adj[u]:
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                if w == 0:
                    dq.appendleft(v)
                else:
                    dq.append(v)

    return [None if d == math.inf else d for d in dist]

File: 05-Algorithms/code/typescript/src/zeroOneBfs.ts
import { AlgoError } from "./errors.js";
import { buildWeightedAdjList, checkVertexInRange, type WeightedEdge } from "./graphCommon.js";

/**
 * 0-1 BFS: shortest paths when every edge weight is 0 or 1, using a deque
 * instead of a priority queue (0-weight edges pushed to the front, 1-weight
 * edges pushed to the back), which is O(V + E) instead of Dijkstra's
 * O(V log V + E). Throws AlgoError("invalid") if any weight is outside
 * {0, 1}. Unreachable vertices are `null`.
 */
export function zeroOneBfs(n: number, edges: WeightedEdge[], src: number, directed = false): (number | null)[] {
  checkVertexInRange(n, src);
  for (const [, , w] of edges) {
    if (w !== 0 && w !== 1) throw new AlgoError("invalid", "zeroOneBfs requires weights in {0, 1}");
  }
  const adj = buildWeightedAdjList(n, edges, directed);
  const dist = new Array(n).fill(Infinity);
  dist[src] = 0;
  const deque: number[] = [src];
  const visited = new Array(n).fill(false);

  while (deque.length > 0) {
    const u = deque.shift() as number;
    if (visited[u]) continue;
    visited[u] = true;
    for (const [v, w] of adj[u]) {
      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w;
        if (w === 0) deque.unshift(v);
        else deque.push(v);
      }
    }
  }

  return dist.map((d) => (d === Infinity ? null : d));
}
